tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -960,42 +945,58 @@ void CommandQueue::enqueue_program(Program& program, bool blocking) {
     // data is cached, then we don't need to stall, otherwise we need to wait for the
     // data to land in DRAM first
     bool stall = false;
-    if (not this->program_to_buffer.count(program_id)) {
+    // No shared cache so far, can come at a later time
+    map<uint64_t, unique_ptr<Buffer>>& program_to_buffer = this->program_to_buffer(this->device->id());
+    if (not program_to_buffer.count(program_id)) {
         stall = true;
-        ProgramMap program_to_device_map = ConstructProgramMap(this->device, program);
+        ProgramMap program_to_device_map = ConstructProgramMap(this->device, program, this->dispatch_core);
 
         vector<uint32_t>& program_pages = program_to_device_map.program_pages;
         uint32_t program_data_size_in_bytes = program_pages.size() * sizeof(uint32_t);
 
         uint32_t write_buffer_command_size = DeviceCommand::NUM_BYTES_IN_DEVICE_COMMAND + program_data_size_in_bytes;
 
-        this->program_to_buffer.emplace(
+        program_to_buffer.emplace(
             program_id,
             std::make_unique<Buffer>(
                 this->device, program_data_size_in_bytes, DeviceCommand::PROGRAM_PAGE_SIZE, BufferType::DRAM));
 
-        this->enqueue_write_buffer(*this->program_to_buffer.at(program_id), program_pages.data(), blocking);
-        this->program_to_dev_map.emplace(program_id, std::move(program_to_device_map));
+        this->enqueue_write_buffer(*program_to_buffer.at(program_id), program_pages.data(), blocking);
+
+        map<uint64_t, ProgramMap>& program_to_dev_map = this->program_to_dev_map(device->id());
+        program_to_dev_map.emplace(program_id, std::move(program_to_device_map));
+
+        const char *READ_BACK_PROGRAMS = std::getenv("TT_METAL_READ_BACK_PROGRAMS");
```

Comment:
did we always have this env var?

Reply:
No, I added it for debug. Can remove if you'd like.